<a href="https://colab.research.google.com/github/organisciak/Scripting-Course/blob/master/labs/05-split-apply-combine-lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Lab 5: Split Apply Combine
Reminder - save your work. Go to File > Save a Copy in Drive to ensure that you have your work saved.



In [ ]:
import pandas as pd
# this bit of code is just for downgrading sqlalchemy, due to issues with the newest version
!pip install -qq sqlalchemy==1.4

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.4/7.4 MB 16.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done


### Movie Dataset Questions

In [ ]:
ratings = pd.read_csv('https://raw.githubusercontent.com/organisciak/Scripting-Course/master/data/movielens_small.csv')
ratings.sample(10)

userId  rating  ...   timestamp  year
19089     654     4.5  ...  1145390078  1989
21525       4     3.0  ...   949896309  1982
68129     213     2.5  ...  1462634437  2014
1848      194     4.0  ...   848098284  1995
22855     202     5.0  ...  1046143639  1986
32289     100     3.0  ...   854193977  1996
89686     564     5.0  ...   974714061  1981
4199       92     3.0  ...   848525694  1995
56398     353     3.5  ...  1157498637  2000
17926     268     4.5  ...  1314894258  1984

[10 rows x 6 columns]

**Questions**

**Q1)** Write the code to determine the median rating by each user. *(4 points)*

In [ ]:
# Answer-Q1


**Q2)** Using groupby, write the code to count how many ratings there are in each genre. *(4 points)*

In [ ]:
# Answer-Q2


**Q3)** The previous question can be answered without groupby. How? *(4 points)*

  *   a) ratings.genres.unique()
  *   b) ratings['genres'].value_counts()
  *   c) ratings[['genres']].unique()








In [ ]:
q3_answer = "" #@param ["", "a) ratings.genres.unique()", "b) ratings['genres'].value_counts()", "c) ratings[['genres']].unique()"]

*Pandas* tip: The answer above is generally a good skill to remember. When you *group by* a single column and your *apply* is `count`, you can also do the function above - which performs a count of all unique values in a column and sorts them.

**Q4)** Which genre has the worst average ratings? *(4 points)*
  *   a) Horror
  *   b) Film-Noir
  *   c) Mystery
  *   d) Sci-Fi









In [ ]:
q4_answer = "" #@param ["", "a) Horror", "b) Film-Noir", "c) Mystery", "d) Sci-Fi"]

**Q5)** Select the three most out-of-fashion film genres, defined by lowest median release year: *(4 points)*
  *   Crime
  *   Documentary
  *   Film-Noir
  *   Musical
  *   Children
  *   Adventure
  *   War



In [ ]:
q5_first = "" #@param ["", 'Crime', 'Documentary', 'Film-Noir', 'Musical', 'Children', 'Adventure', 'War']
q5_second = "" #@param ["", 'Crime', 'Documentary', 'Film-Noir', 'Musical', 'Children', 'Adventure', 'War']
q5_third = "" #@param ["", 'Crime', 'Documentary', 'Film-Noir', 'Musical', 'Children', 'Adventure', 'War']
q5_answer = sorted([q5_first, q5_second, q5_third])

**Q6)** Which post-2011 film has the most deviation among ratings? *(4 points)*
    
*   The Heat
*   Twilight Saga: Breaking Dawn
*   The Secret Life of Pets
*  Drinking Buddies

In [ ]:
q6_answer = "" #@param ["", "a) The Heat", "b) Twilight Saga: Breaking Dawn", "c) The Secret Life of Pets", "d) Driniking Buddies"]

**Q7)** Get all the unique combinations of title+year. Submit the code. *(4 points)*

In [ ]:
# Answer-Q7


**Q8)** Run `ratings['movie']`. What does the error mean? *(3 points)*

In [ ]:
q8_answer = "" #@param {type:"string"}

**Q9)** What does the error message mean for the following code: `ratings['new_column'] = ratings[ratings.genres == 'Drama']` ? *(4 points)*

In [ ]:
q9_answer = "" #@param {type:"string"}

**Q10)** Find all movie titles that are used by multiple movies. Consider a unique 'movie' as a unique title+year pair. Share your code. *(8 points)*

(Hint - there should be 207 movie titles that you return).

In [ ]:
# Answer-Q10


## Input / Output

**Q11)** For the movie data: write the code to save a CSV of year/title/mean_rating. Save it in the default folder and call it `mean_ratings.csv`. *(5 points)*

In [ ]:
# Answer-Q11


In [ ]:
#@markdown **Q12)** Write the code to open our book circulation database from two weeks ago, and retrieve information on patrons that are at least 18 years old as a DataFrame. *(7 points)*

#@markdown *Run this cell to download our circulation data sqlite database again*
!wget -nv https://github.com/organisciak/Scripting-Course/blob/master/data/circulation.sqlite?raw=true -O circulation.sqlite

#@markdown Tip - your database connection string will be `'sqlite:///circulation.sqlite'`

In [ ]:
# Answer-Q12


**Q13)** For patrons 18 years old and older, what is the most common home city? *(5 points)*

*   Thomasmouth
*   New Hannah
*   Michelefurt

*Hint - the second most common city in your results should be a 3-way tie, between 'East Brookefort' and two other cities.*



In [ ]:
q13_answer = "" #@param ["", "a) Thomasmouth", "b) New Hannah", "c) Michelefurt"]

**Q14)** Write the code to determine the author with the most books checked out. Tip: remember joins! You can write a query for Pandas to retrieve a single, already joined table (you know how to do this). If you're adventurous and want to try an alternative that we didn't learn, it is also possible to do the join in Pandas. *(10 points)*

*Hint*: The answer is J.K. Rowling with 63 books.

In [ ]:
# Answer-Q14


**Q15.** 

a) Write the code to retrieve the list of main songs recorded by the Beatles from Wikipedia. *(8 points)*

*Hint*: The answer should have 213 rows (or close - Wikipedia is ever-changing!).

In [ ]:
# Answer-Q15a


In [ ]:
#@markdown **b)** How many songs by Carl Perkins did the Beatles record? (*3pts*)

q15b_answer = 0 #@param{type:'integer'}

#@markdown **c)** What's the median year of a unique Beatles recording (i.e. the year where half of the songs were recorded before, and half after)? (*3pts*)

q15c_answer = 0 #@param{type:'integer'}

#@markdown **d)** From the typical songwriting combinations in the group (LennonMcCartney, Harrison, and Starkey), which one had the latest mean songwriting year?^ (*4pts*)
q15d_answer = '' #@param ["", "LennonMcCartney", "Harrison", "Starkey"]

## Is In

Not all selection is done with logical operators. One useful method for a `Series` is `isin()`.

`isin()` takes a list or collection of values, and returns True/False for whether each item in the Series is in that set of values. Remember that DataFrame columns are Series'.

Example: Find movies where the genre is one of the following: Drama, Action, Sci-Fi.

In [ ]:
sample = ratings.sample(10)
sample

userId  rating  ...   timestamp  year
21514     509     3.0  ...   940642681  1979
90017     529     1.0  ...   959975472  1957
85306     192     2.0  ...   843634385  1994
55865     311     4.0  ...  1094585477  2000
9171      150     5.0  ...  1113907633  1994
8755      435     4.0  ...   891508567  1991
92537     190     3.0  ...   975694648  1999
26541     541     4.0  ...   976829924  1999
86358     187     3.0  ...  1241388708  1990
91129     132     2.5  ...  1283580984  1999

[10 rows x 6 columns]

In [ ]:
sample.genres.isin(['Action', 'Drama', 'Sci-Fi'])

21514    False
90017     True
85306    False
55865     True
9171     False
8755     False
92537    False
26541    False
86358    False
91129     True
Name: genres, dtype: bool

In [ ]:
sample[sample.genres.isin(['Action', 'Drama', 'Sci-Fi'])]

userId  rating  ...   timestamp  year
90017     529     1.0  ...   959975472  1957
55865     311     4.0  ...  1094585477  2000
91129     132     2.5  ...  1283580984  1999

[3 rows x 6 columns]

***Q16)***

Practice using `isin()` yourself, then try the following bigger question:
    
- Write the code to select all ratings by the 10 most prolific users. It's fine if you manually write the list of prolific users' ids - but even so, you'll need to figure out how to get them! (*12 pts*)

Common error: remember that you're processing *all* ratings, not the 10 row `sample` DataFrame.

In [ ]:
# Answer-Q16


# Submission Instructions

In [ ]:
#@markdown ### First, Enter your name for grading
my_name = "" #@param { type:'string' }

#@markdown _Have you saved your work for yourself? Don't forget to Save a Copy in Drive so that you have your progress._

In [ ]:
#@markdown ### Second, check your work:

#@markdown - have you answered all the questions?
#@markdown - Does this notebook run from top to bottom?
#@markdown     - Go to "Runtime > Restart and run all..." to check. Do all the cells run, to the very bottom, or is there a cell in the middle with an error?
#@markdown - Have you completed all the answers where you entered code, keeping the `# Answer-Qx` line at the start of those cells?

#@markdown *A lab that the professor has to fix manually will lose 10pts - run the checks!*

#@markdown ### Finally, submit it

#@markdown - Download the file with "File > Download .ipynb" and submit it to the Canvas assignment page.